components/brave_news/browser/direct_feed_fetcher.cc

```diff
@@ -143,15 +146,55 @@ DirectFeedResponse::~DirectFeedResponse() = default;
 DirectFeedResponse::DirectFeedResponse(DirectFeedResponse&&) = default;
 
 DirectFeedFetcher::DirectFeedFetcher(
-    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
-    : url_loader_factory_(url_loader_factory) {}
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    base::WeakPtr<Delegate> delegate)
+    : url_loader_factory_(url_loader_factory), delegate_(delegate) {}
 DirectFeedFetcher::~DirectFeedFetcher() = default;
 
-void DirectFeedFetcher::DownloadFeed(const GURL& url,
+void DirectFeedFetcher::DownloadFeed(GURL url,
                                      std::string publisher_id,
                                      DownloadFeedCallback callback) {
+  if (url.SchemeIs(url::kHttpScheme)) {
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE,
+        base::BindOnce(
+            [](base::WeakPtr<Delegate> delegate,
+               scoped_refptr<base::SingleThreadTaskRunner> source_task_runner,
+               base::OnceCallback<void(bool)> callback, GURL url) {
+              bool should_upgrade = delegate->ShouldUpgradeToHttps(url);
```

Comment:
I think you need to check |delegate| here before using it

Reply:
isn't `delegate` guaranteed to be valid within the callback? 

> A callback bound using a weak pointer receiver will be automatically cancelled (calling Run() becomes a no-op) if the weak pointer is invalidated, e.g. its associated class instance is destroyed.

https://source.chromium.org/chromium/chromium/src/+/main:docs/callback.md

Reply:
> isn't `delegate` guaranteed to be valid within the callback?
> 
> > A callback bound using a weak pointer receiver will be automatically cancelled (calling Run() becomes a no-op) if the weak pointer is invalidated, e.g. its associated class instance is destroyed.
> 
> https://source.chromium.org/chromium/chromium/src/+/main:docs/callback.md

It's auto-cancelled only if you use it as a pointer to a class/struct method. If you have a function that receives `WeakPtr` explicitly, you should check it manually:

```
// auto-cancelled if weakptr is invalidated
base::BindOnce(&MyClass::Func, my_class_.GetWeakPtr());

// will be called even if weakptr is invalidated
base::BindOnce([](base::WeakPtr<MyClass> c), my_class_.GetWeakPtr());
```

Reply:
hmm, interesting. will add the check then